shell/src/main/java/alluxio/cli/docgen/ConfigurationDocValidator.java

```diff
@@ -0,0 +1,176 @@
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utility to validate existence of any necessary configuration docGen changes during pull request.
+ */
+@ThreadSafe
+@PublicApi
+public final class ConfigurationDocValidator extends DocValidator {
+  private static final Logger LOG = LoggerFactory.getLogger(ConfigurationDocValidator.class);
+  public static final String CSV_FILE_HEADER = "propertyName,defaultValue";
+
+  private ConfigurationDocValidator() {
+  } // prevent instantiation
+
+  /**
+   * Validates configuration files.
+   *
+   * @return returns number of deltas
+   */
+  public static Integer validate() throws IOException {
+    Collection<? extends PropertyKey> defaultKeys = PropertyKey.defaultKeys();
+    defaultKeys.removeIf(PropertyKey::isHidden);
+
+    // Sort defaultKeys
+    List<PropertyKey> dfkeys = new ArrayList<>(defaultKeys);
+    Collections.sort(dfkeys);
+
+    Integer numDeltas = 0;
+    numDeltas += validateFiles(dfkeys, CSV_FILE_DIR, CSV_FILE_NAMES);
+    numDeltas += validateFiles(dfkeys, YML_FILE_DIR, YML_FILE_NAMES);
+
+    return numDeltas;
+  }
+
+  private static Integer validateFiles(List<PropertyKey> dfkeys,
+                                       String fileDir, List<String> fileNames)
+          throws IOException {
+
+    String filePath = PathUtils.concatPath(Configuration.getString(PropertyKey.HOME), fileDir);
+    Integer numDeltas = 0;
+    for (String fileName : fileNames) {
+      String key = fileName.substring(0, fileName.indexOf("configuration") - 1);
+
+      // based on key substring within each filename, create a regex to
+      // filter for specific property keys.
+      List<String> rPatterns = new ArrayList<String>();
+      switch (key) {
+        case "user":
+          rPatterns.add("^alluxio\\.(user|user\\..*)");
+          break;
+        case "master":
+          rPatterns.add("^alluxio\\.(master|master\\..*)");
+          break;
+        case "worker":
+          rPatterns.add("^alluxio\\.(worker|worker\\..*)");
+          break;
+        case "security":
+          rPatterns.add("^alluxio\\.(security|security\\..*)");
+          break;
+        case "cluster-management":
+          rPatterns.add("^alluxio\\.(integration|integration\\..*)");
+          break;
+        case "common":
+          rPatterns.add("^alluxio\\.(?!(user|master|worker|security|integration)\\b)\\w*[\\.]?.*");
+          // additional regex pattern to retain non-system configuration records.
+          rPatterns.add("^(fs|s3a)\\b\\..*");
+          break;
+        default:
+          // ignore and skip any non-relevant files.
+          LOG.warn("Ignoring unrecognized filename substring key: " + key);
+          continue;
+      }
+
+      List<Pattern> compiledRegex = new ArrayList<Pattern>();
+      for (String pattern : rPatterns) {
+        compiledRegex.add(Pattern.compile(pattern));
+      }
+
+      List<String> revised = new ArrayList<String>();
+      if (fileDir.equals(CSV_FILE_DIR)) {
```

Comment:
the validation function is handling both yaml and csv files, yet there is a separate generation function for each. this instinctively seems very wrong. not only are you rewriting the same logic in two different places (ex. add csv file header in both the validation and generation -> this means if we were to remove the header, it needs to be removed in two places), but there aren't mirrored codepaths when comparing verification vs generation.

these are the approaches i'm imagining could've happened...

best case -> 1 generalized function to handle generation and validation. everything is properly refactored to maintain the principle of single source of truth for any piece of logic

acceptable case, based on the fact that csv and yaml generation logic is currently separated -> incorporate the verification logic such that the unique logic between csv vs yaml is not duplicated. i'm assuming that there is a good reason to separate them (or maybe my faith is misplaced on the previous committer/reviewer)

copy-paste everything case -> 2x2 set of functions for handling each of the 4 permutations of generation types vs execution types, assuming there is unique logic in each of the cases so it's not advisable to combine (though it is doubtful that this assumption is true)

inconsistent case -> separate functions for one permutation set, combined function for the other permutation set that contains the unique logic of the permutations it has to handle, but repeated from the separated functions

i'm not sure which one is worse, the inconsistent case or the copy and paste case, but neither should be acceptable.